### ISSUE TITLE
react - using CSS Modules in a Nrwl NX workspace

### ISSUE THREAD

**Issue by maku** [nrwl/nx#1744]
I run version 8.4.12 of nx

I would like use css modules in a react app 

In a recent react project, created with "create-react-app", I am able to use css modules.

e.g.
 import `classes from './MyComponent.module.css';`

and the usage:

`<div className={classes.my-css-class}> </div>`

How is this possible with nx?

**Comment by jaysoo** [comment]
Hi @maku 

We don't currently support CSS modules but we can look at extending our builder for it.

The workaround right now is to provide your own custom webpack to the web builder.

Look in your `workspace.json` file for your app and you should find a `build` entry under `architect`.

```
  "build": {
    "builder": "@nrwl/web:build",
    "options": {
      "differentialLoading": false,
      "outputPath": "dist/apps/demo",
      "index": "apps/demo/src/index.html",
      "main": "apps/demo/src/main.tsx",
      "polyfills": "apps/demo/src/polyfills.ts",
      "tsConfig": "apps/demo/tsconfig.app.json",
      "assets": ["apps/demo/src/favicon.ico", "apps/demo/src/assets"],
      "styles": ["apps/demo/src/styles.css"],
      "scripts": [],
      "webpackConfig": "@nrwl/react/plugins/babel"
    }
```

You need to switch `webpackConfig` with your own, such as `"./apps/demo/webpack.config.js"`. Please change `demo` to your actual app name.

Now, you need to provide the config function like this:

https://gist.github.com/jaysoo/431828f47e5202d3e4afc6c34fdac2c0

This function modifies the webpack config by adding a rule to handle `.module.css` files, as well as exclude these files from other rules.

**Comment by maku** [comment]
Hi @jaysoo, 

thank you for answering. I would only use css modules when I don't have to fiddle around with the config (mainly I use Angular but I am curious about the React features in nx). But from my point of view it would make sense when nx is able to handle css modules...

**Comment by jaysoo** [comment]
@maku Sounds good. I'm doing some refactoring on our babel support. After that is done this should be easy to add.

**Comment by sanjevirau** [comment]
@jaysoo I tried configuring the webpack configuration like mentioned above. Nx compiled it successfully but the class name doesn't seems to be shown in the render. Am I missing something?

**Comment by jaysoo** [comment]
@sanjevirau See this demo here https://github.com/jaysoo/css-modules-example. Make sure you change `webpackConfig` in your `workspace.json` (or `angular.json` if you are using Angular CLI).

**Comment by jaysoo** [comment]
Support for CSS modules have been added. It's the same as CRA so you need to name your files as `*.module.css` in order for it to kick in.

**Comment by LucaFilitz** [comment]
How can I use scss modules?

**Comment by TheAifam5** [comment]
Same here, but for less?

**Comment by jaysoo** [comment]
@LucaFilitz @TheAifam5 You will need to provide your own webpack config extensions.

Something like this:

- Switch out the default `webpackConfig` path
    * https://github.com/jaysoo/css-modules-example/blob/master/workspace.json#L22
- Add a rule for CSS Modules + LESS/SCSS/...
    * https://github.com/jaysoo/css-modules-example/blob/master/apps/demo/webpack.config.js#L21-L34

**Comment by LucaFilitz** [comment]
`const merge = require(`webpack-merge`);
const getReactWebpackConfig = require(`@nrwl/react/plugins/webpack`);
const autoprefixer = require(`autoprefixer`);

module.exports = config => {
  const original = getReactWebpackConfig(config);
  return merge({
    customizeArray: merge.unique(`module.rules`, [`ScssLoader`], rule =>
      rule.test.test(`example.scss`) ? `ScssLoader` : ``
    )
  })(
    {
      module: {
        rules: [
          {
            test: /\.module\.scss$/,
            use: [
              `style-loader`,
              {
                loader: `css-loader`,
                options: {
                  modules: true,
                  importLoaders: 1
                }
              },
              {
                loader: `postcss-loader`,
                options: {
                  ident: `postcss`,
                  sourceMap: true,
                  plugins: () => [autoprefixer()]
                }
              },
              `sass-loader`
            ]
          },
          {
            test: /^((?!module).)*\.s[ac]ss$/,
            use: [`style-loader`, `css-loader`, `sass-loader`]
          },
          {
            test: /\.css$/,
            use: [`style-loader`, `css-loader`]
          }
        ]
      }
    },
    original
  );
};
`
For other people
This is a custom webpack config which support .css, .s[a/c]ss and .module.scss
You could easily add .module.css if you want.
Hope this helps someone :)

**Comment by LucaFilitz** [comment]
https://hastebin.com/jugohuxupu.js

This is a custom webpack config for .s[a/c]ss, .css and .module.s[a/c]ss
If you want you can easily add .module.css
Hope this helps anyone :)

**Comment by maierson** [comment]
This works well except when adding sass variables:
```
// scss
$box-color: #fff;

.my-box {
   background-color: $box-color;
}
```

Which doesn't replace the variable with its value in the compiled css
```
// css
.hash {
  background-color: $box-color; // this should be #fff
}
```
(it does work with non module .scss files out of the box).

**Comment by katyapavlenko** [comment]
@jaysoo thank you for css mosules support!
But I see that initial class names is removed in serve mode
I really miss them for debugging. Is there any settings I can change to see initial class names?
<img width="326" alt="Screen Shot 2021-01-31 at 13 34 04" src="https://user-images.githubusercontent.com/55044564/106382548-fdf17180-63c8-11eb-8de9-ddb06ce414fd.png">

**Comment by itsTalwar** [comment]
@jaysoo this works well for apps, but it is failing in libs.

**Comment by KamalAman** [comment]
Seems to have been fixed in https://github.com/nrwl/nx/pull/5479

**Comment by 20Koen02** [comment]
Still failing in libs? Any work arounds?

**Comment by cjclifford** [comment]
Is there a reason that SCSS modules are used by default if they're not supported?

**Comment by solshark** [comment]
Why this is closed?

**Comment by lihongwang** [comment]
Any update about lib support SCSS modules?

**Comment by niyowski** [comment]
I get the following error for a React component in a `ui` lib while using CSS modules:

> Property 'container' comes from an index signature, so it must be accessed with ['container']

I disabled [noPropertyAccessFromIndexSignature](https://stackoverflow.com/a/70748402/19377445). The error in IDE gone but the app didn't run.

**Comment by rishabh2712** [comment]
The problem still persists. The IDE config is gone, but the dev build still fails.

**Comment by espretto** [comment]
If `*.css` files where to be handled by a custom loader, that would have to be configured in both the webpack and rollup configs using their respective plugins.

**Comment by Zuravvski** [comment]
> @jaysoo this works well for apps, but it is failing in libs.

TL;DR: you have at least 2 ways to fix that.

1. Switch to `babel-jest` in your lib's `jest.confg.js` like so:
```js
module.exports = {
  displayName: 'shared-core',
  preset: '../../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest', // <----- the default is 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/shared/core',
};
```
2. In case you want to stick with `ts-jest` you have to add type declarations to your lib's `tsconfig.spec.json`
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/out-tsc",
    "module": "commonjs",
    "types": ["jest", "node"]
  },
  "include": [
    "**/*.test.ts",
    "**/*.spec.ts",
    "**/*.test.tsx",
    "**/*.spec.tsx",
    "**/*.test.js",
    "**/*.spec.js",
    "**/*.test.jsx",
    "**/*.spec.jsx",
    "**/*.d.ts"
  ]
 // Below file declaration is what you need to add. Make sure that your path is correct
  "files": [
    "../../../node_modules/@nrwl/react/typings/cssmodule.d.ts"
  ],
}
```
---
### Full solution
I'm going to include full configuration that handles SVG components as well, so you don't have to come back here with yet another issue.

1. Since each `jest.config.js` uses the same `jest.preset.js` preset file we can implement the module name mappers and transforms in one place. 

**jest.preset.js**
```
const nxPreset = require('@nrwl/jest/preset');

module.exports = { 
    ...nxPreset,
    transform: {
        ...nxPreset.transform, // we want to keep the original transforms
        '\\.(jpg|jpeg|png|gif|ico|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga|webmanifest|xml)$':
            '__dirname/node_modules/@nrwl/react/plugins/jest.js' // SVG transformer
   },
  moduleNameMapper: {
        '\\.(css)$': 'identity-obj-proxy'
  },
};
```

Regexes have been copied over from minimal jest config provided in VS Code's jest runner docs:
https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner

As for the SVG transformer, it seems that one is provided in `@nrwl/react` itself, but for some reason is not used by default.
https://github.com/nrwl/nx/tree/master/packages/react/plugins

2. If you want to stick to 'ts-jest' you would have to adjust `tsconfig.spec.json` as I depicted up above in the `TL;DR` section
3. If SVG components still fail your tests, you have to extend the webpack.config.js with [webpack plugin](https://github.com/nrwl/nx/blob/master/packages/react/plugins/webpack.ts) manually

---
### Suggested @nrwl/react changes
- Maybe we could have custom `jest.preset.js` in @nrwl/react that extends the one from `@nrwl/jest` with the plugins that you provide in the package already.
- `tsconfig.spec.json` should have a custom template in the [library generator](https://github.com/nrwl/nx/tree/master/packages/react/src/generators/library) to fill the `files` section with relevant types as you do in  [tsconfig.lib.json](https://github.com/nrwl/nx/blob/master/packages/react/src/generators/library/files/lib/tsconfig.lib.json__tmpl__)
